packages/jsii-rosetta/lib/snippet-dependencies.ts

```diff
@@ -0,0 +1,245 @@
+      );
+    }
+
+    return a;
+  }
+
+  if (a.type === 'symbolic' && b.type === 'concrete') {
+    // Reverse roles so we fall into the previous case
+    return resolveConflict(name, b, a);
+  }
+
+  throw new Error('Cases should have been exhaustive');
+}
+
+/**
+ * Check that the directory we were given has all the necessary dependencies in it
+ *
+ * It's a warning if this is not true, not an error.
+ */
+export async function validateAvailableDependencies(directory: string, deps: Record<string, CompilationDependency>) {
+  const failures = await Promise.all(
+    Object.entries(deps).flatMap(async ([name, _dep]) => {
+      try {
+        await findDependencyDirectory(name, directory);
+        return [];
+      } catch {
+        return [name];
+      }
+    }),
+  );
+
+  if (failures.length > 0) {
+    logging.warn(
+      `${directory}: packages necessary to compile examples missing from supplied directory: ${failures.join(', ')}`,
+    );
+  }
+}
+
+/**
+ * Prepare a temporary directory with symlinks to all the dependencies we need.
+ *
+ * - Symlinks the concrete dependencies
+ * - Tries to first find the symbolic dependencies in a potential monorepo that might be present
+ *   (try both `lerna` and `yarn` monorepos).
+ * - Installs the remaining symbolic dependencies using 'npm'.
+ */
+export async function prepareDependencyDirectory(deps: Record<string, CompilationDependency>): Promise<string> {
+  const concreteDirs = Object.values(deps)
+    .filter(isConcrete)
+    .map((x) => x.resolvedDirectory);
+  const monorepoPackages = await scanMonoRepos(concreteDirs);
+
+  const tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'rosetta'));
+  logging.info(`Preparing dependency closure at ${tmpDir}`);
+
+  // Resolved symbolic packages against monorepo
+  const resolvedDeps = mkDict(
+    Object.entries(deps).map(([name, dep]) => [
+      name,
+      dep.type === 'concrete'
+        ? dep
+        : ((monorepoPackages[name]
+            ? { type: 'concrete', resolvedDirectory: monorepoPackages[name] }
+            : dep) as CompilationDependency),
+    ]),
+  );
+
+  // Use 'npm install' only for the symbolic packages. For the concrete packages,
+  // npm is going to try and find transitive dependencies as well and it won't know
+  // about monorepos.
+  const symbolicInstalls = Object.entries(resolvedDeps).flatMap(([name, dep]) =>
+    isSymbolic(dep) ? [`${name}@${dep.versionRange}`] : [],
+  );
+  const linkedInstalls = mkDict(
+    Object.entries(resolvedDeps).flatMap(([name, dep]) =>
+      isConcrete(dep) ? [[name, dep.resolvedDirectory] as const] : [],
+    ),
+  );
+
+  // Run 'npm install' on it
+  await fs.mkdirp(tmpDir);
+
+  if (symbolicInstalls.length > 0) {
+    logging.debug(`Installing example dependencies: ${symbolicInstalls.join(' ')}`);
+    cp.execSync(`npm install ${symbolicInstalls.join(' ')}`, { cwd: tmpDir, encoding: 'utf-8' });
```

Comment:
Perhaps also turn on "global style" install? This will prevent the package's own dependencies from being installed/de-duped at the same level, too, so you guarantee the list of directories under `node_modules` after this is done is exactly the list of package names you passed.

That might however have unintended consequences (i.e: multiple different versions of transitive dependencies), which might be problematic (I'm not sure).

Reply:
I don't think this is an issue. The packages should all be able to be installed into a single flat closure.